File: app/src/main/java/org/ole/planet/myplanet/ui/enterprises/EnterpriseCalendarFragment.java
package org.ole.planet.myplanet.ui.enterprises;


import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.DatePicker;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import com.events.calendar.views.EventsCalendar;
import com.google.gson.Gson;
import com.google.gson.JsonObject;

import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.model.RealmMeetup;
import org.ole.planet.myplanet.model.RealmMyTeam;
import org.ole.planet.myplanet.ui.team.BaseTeamFragment;
import org.ole.planet.myplanet.utilities.TimeUtils;
import org.ole.planet.myplanet.utilities.Utilities;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.UUID;

/**
 * A simple {@link Fragment} subclass.
 */
public class EnterpriseCalendarFragment extends BaseTeamFragment {

    EventsCalendar calendarView;
    List<RealmMeetup> list;
    TextView startDate, startTime, endDate, endTime;
    Calendar start, end;

    public EnterpriseCalendarFragment() {
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_enterprise_calendar, container, false);
        start = Calendar.getInstance();
        end = Calendar.getInstance();
        calendarView = v.findViewById(R.id.calendar);
        v.findViewById(R.id.add_event).setOnClickListener(view -> showMeetupAlert());
        return v;
    }

    private void showMeetupAlert() {
        View v = LayoutInflater.from(getActivity()).inflate(R.layout.add_meetup, null);
        TextView title = v.findViewById(R.id.et_title);
        TextView location = v.findViewById(R.id.et_location);
        TextView description = v.findViewById(R.id.et_description);
        RadioGroup radioGroup = v.findViewById(R.id.rg_recuring);
        startDate = v.findViewById(R.id.tv_start_date);
        startTime = v.findViewById(R.id.tv_start_time);
        endDate = v.findViewById(R.id.tv_end_date);
        endTime = v.findViewById(R.id.tv_end_time);
        setDatePickerListener(startDate, start);
        setDatePickerListener(endDate, end);
        setTimePicker(startTime);
        setTimePicker(endTime);

        new AlertDialog.Builder(getActivity()).setView(v)
                .setPositiveButton("Save", (dialogInterface, i) -> {
                    String ttl = title.getText().toString();
                    String desc = description.getText().toString();
                    String loc = location.getText().toString();
                    if (ttl.isEmpty()) {
                        Utilities.toast(getActivity(), "Title is required");
                    } else if (desc.isEmpty()) {
                        Utilities.toast(getActivity(), "Description is required");
                    } else if (start == null) {
                        Utilities.toast(getActivity(), "Start time is required");
                    } else {
                        if (!mRealm.isInTransaction())
                            mRealm.beginTransaction();
                        RealmMeetup meetup = mRealm.createObject(RealmMeetup.class, UUID.randomUUID().toString());
                        meetup.setTitle(ttl);
                        meetup.setDescription(desc);
                        meetup.setMeetupLocation(loc);
                        meetup.setCreator(user.getId());
                        meetup.setStartDate(start.getTimeInMillis());
                        if (end != null)
                            meetup.setEndDate(end.getTimeInMillis());
                        meetup.setEndTime(endTime.getText().toString());
                        meetup.setStartTime(startTime.getText().toString());
                        RadioButton rb = v.findViewById(radioGroup.getCheckedRadioButtonId());
                        if (rb != null) {
                            meetup.setRecurring(rb.getText().toString());
                        }
                        JsonObject ob = new JsonObject();
                        ob.addProperty("teams", teamId);
                        meetup.setLinks(new Gson().toJson(ob));
                        meetup.setTeamId(teamId);
                        mRealm.commitTransaction();
                        Utilities.toast(getActivity(), "Meetup added");
                    }
                }).setNegativeButton("Cancel", null).show();
    }


    private void setDatePickerListener(TextView view, Calendar date) {
        Calendar c = Calendar.getInstance();
        view.setOnClickListener(v -> {

            new DatePickerDialog(getActivity(), (vi, year, monthOfYear, dayOfMonth) -> {
                date.set(Calendar.YEAR, year);
                date.set(Calendar.MONTH, monthOfYear);
                date.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                if (view != null)
                    view.setText(TimeUtils.formatDate(date.getTimeInMillis(), "yyyy-MM-dd"));
            }, c.get(Calendar.YEAR), c.get(Calendar.MONTH), c.get(Calendar.DAY_OF_MONTH)).show();

        });
    }


    private void setTimePicker(TextView time) {
        Calendar c = Calendar.getInstance();
        time.setOnClickListener(v -> {
            TimePickerDialog timePickerDialog = new TimePickerDialog(getActivity(),
                    (view, hourOfDay, minute) -> {
                        time.setText(String.format("%02d:%02d", hourOfDay, minute));
                    }, c.get(Calendar.HOUR_OF_DAY), c.get(Calendar.MINUTE), true);
            timePickerDialog.show();
        });

    }


    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        list = mRealm.where(RealmMeetup.class).equalTo("teamId", teamId).findAll();
        Utilities.log(list.size() + "");
        Calendar daySelected = Calendar.getInstance();
        daySelected.add(Calendar.DAY_OF_MONTH, 5);
        List<Calendar> li = new ArrayList<>();
        for (RealmMeetup meetup : list
        ) {
            Calendar c = Calendar.getInstance();
            c.setTimeInMillis(meetup.getStartDate());
        }
    }


}

File: app/src/main/java/org/ole/planet/myplanet/ui/enterprises/EnterpriseCalendarFragment.kt
package org.ole.planet.myplanet.ui.enterprises

import android.app.*
import android.content.Context
import android.content.res.Resources
import android.os.*
import android.util.TypedValue
import android.view.*
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.applandeo.materialcalendarview.CalendarDay
import com.applandeo.materialcalendarview.CalendarView
import com.applandeo.materialcalendarview.listeners.OnCalendarDayClickListener
import com.google.gson.*
import io.realm.Realm
import io.realm.RealmResults
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.databinding.*
import org.ole.planet.myplanet.model.RealmMeetup
import org.ole.planet.myplanet.model.RealmNews
import org.ole.planet.myplanet.service.UploadManager
import org.ole.planet.myplanet.ui.mymeetup.AdapterMeetup
import org.ole.planet.myplanet.ui.team.BaseTeamFragment
import org.ole.planet.myplanet.utilities.*
import java.text.SimpleDateFormat
import java.time.Instant
import java.time.ZoneId
import java.util.*

class EnterpriseCalendarFragment : BaseTeamFragment() {
    private lateinit var fragmentEnterpriseCalendarBinding: FragmentEnterpriseCalendarBinding
    private val selectedDates: MutableList<Calendar> = mutableListOf()
    private lateinit var calendar: CalendarView
    private lateinit var list: List<Calendar>
    private lateinit var start: Calendar
    private lateinit var end: Calendar
    private lateinit var clickedCalendar: Calendar
    private lateinit var calendarEventsMap: MutableMap<CalendarDay, RealmMeetup>
    private lateinit var meetupList: RealmResults<RealmMeetup>
    private val eventDates: MutableList<Calendar> = mutableListOf()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        fragmentEnterpriseCalendarBinding = FragmentEnterpriseCalendarBinding.inflate(inflater, container, false)
        start = Calendar.getInstance()
        end = Calendar.getInstance()
        return fragmentEnterpriseCalendarBinding.root
    }

    private fun showMeetupAlert() {
        val addMeetupBinding = AddMeetupBinding.inflate(layoutInflater)
        setDatePickerListener(addMeetupBinding.tvStartDate, start, end)
        setDatePickerListener(addMeetupBinding.tvEndDate, end, null)
        setTimePicker(addMeetupBinding.tvStartTime)
        setTimePicker(addMeetupBinding.tvEndTime)
        if (!::clickedCalendar.isInitialized) {
            clickedCalendar = Calendar.getInstance()
        }

        val alertDialog = AlertDialog.Builder(requireActivity()).setView(addMeetupBinding.root)
            .setPositiveButton("Save") { _, _ ->
                val title = "${addMeetupBinding.etTitle.text}"
                val description = "${addMeetupBinding.etDescription.text}"
                val location = "${addMeetupBinding.etLocation.text}"
                if (title.isEmpty()) {
                    Utilities.toast(activity, getString(R.string.title_is_required))
                } else if (description.isEmpty()) {
                    Utilities.toast(activity, getString(R.string.description_is_required))
                } else {
                    try {
                        if (!mRealm.isInTransaction) {
                            mRealm.beginTransaction()
                        }
                        val meetup = mRealm.createObject(RealmMeetup::class.java, "${UUID.randomUUID()}")
                        meetup.title = title
                        meetup.description = description
                        meetup.meetupLocation = location
                        meetup.creator = user?.name
                        meetup.startDate = start.timeInMillis
                        meetup.endDate = end.timeInMillis
                        meetup.endTime = "${addMeetupBinding.tvEndTime.text}"
                        meetup.startTime = "${addMeetupBinding.tvStartTime.text}"
                        meetup.createdDate = System.currentTimeMillis()
                        meetup.sourcePlanet = team?.teamPlanetCode
                        val jo = JsonObject()
                        jo.addProperty("type", "local")
                        jo.addProperty("planetCode", team?.teamPlanetCode)
                        meetup.sync = Gson().toJson(jo)
                        val rb = addMeetupBinding.rgRecuring.findViewById<RadioButton>(addMeetupBinding.rgRecuring.checkedRadioButtonId)
                        if (rb != null) {
                            meetup.recurring = "${rb.text}"
                        }
                        val ob = JsonObject()
                        ob.addProperty("teams", teamId)
                        meetup.link = Gson().toJson(ob)
                        meetup.teamId = teamId
                        mRealm.commitTransaction()
                        Utilities.toast(activity, getString(R.string.meetup_added))
                        refreshCalendarView()
                    } catch (e: Exception) {
                        mRealm.cancelTransaction()
                        e.printStackTrace()
                        Utilities.toast(activity, getString(R.string.meetup_not_added))
                    }
                }
            }.setNegativeButton("Cancel", null).create()

        alertDialog.setOnDismissListener {
            if (selectedDates.contains(clickedCalendar)) {
                selectedDates.remove(clickedCalendar)
                refreshCalendarView()
            }
        }
        alertDialog.show()
        alertDialog.window?.setBackgroundDrawableResource(R.color.card_bg)
    }

    private fun setDatePickerListener(view: TextView, date: Calendar?, endDate: Calendar?) {
        val c = Calendar.getInstance()
        if (date != null && endDate != null) {
            view.text = date.timeInMillis.let { it1 -> TimeUtils.formatDate(it1, "yyyy-MM-dd") }
        }
        view.setOnClickListener {
            DatePickerDialog(requireActivity(), { _, year, monthOfYear, dayOfMonth ->
                date?.set(Calendar.YEAR, year)
                date?.set(Calendar.MONTH, monthOfYear)
                date?.set(Calendar.DAY_OF_MONTH, dayOfMonth)
                if(endDate != null && date == Calendar.getInstance()){
                    endDate.set(Calendar.YEAR, year)
                    endDate.set(Calendar.MONTH, monthOfYear)
                    endDate.set(Calendar.DAY_OF_MONTH, dayOfMonth)
                }
                view.text = date?.timeInMillis?.let { it1 -> TimeUtils.formatDate(it1, "yyyy-MM-dd") }
            }, c.get(Calendar.YEAR), c.get(Calendar.MONTH), c.get(Calendar.DAY_OF_MONTH)).show()
        }
    }

    private fun setTimePicker(time: TextView) {
        val c = Calendar.getInstance()
        time.setOnClickListener {
            val timePickerDialog = TimePickerDialog(
                activity, { _, hourOfDay, minute ->
                    time.text = String.format(Locale.getDefault(), "%02d:%02d", hourOfDay, minute) },
                c.get(Calendar.HOUR_OF_DAY), c.get(Calendar.MINUTE), true)
            timePickerDialog.show()
        }
    }

    override fun onResume() {
        super.onResume()
        setupCalendarClickListener()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        list = mutableListOf()
        calendar = fragmentEnterpriseCalendarBinding.calendarView
        calendarEventsMap = mutableMapOf()
        setupCalendarClickListener()
    }

    private fun setupCalendarClickListener(){
        fragmentEnterpriseCalendarBinding.calendarView.setOnCalendarDayClickListener(object : OnCalendarDayClickListener {
            override fun onClick(calendarDay: CalendarDay) {
                meetupList = mRealm.where(RealmMeetup::class.java).equalTo("teamId", teamId).findAll()
                clickedCalendar = calendarDay.calendar
                val clickedDateInMillis = clickedCalendar.timeInMillis
                val clickedDate = Instant.ofEpochMilli(clickedDateInMillis)
                    .atZone(ZoneId.systemDefault())
                    .toLocalDate()

                val markedDates = meetupList.mapNotNull { meetup ->
                    val meetupDate = Instant.ofEpochMilli(meetup.startDate)
                        .atZone(ZoneId.systemDefault())
                        .toLocalDate()
                    if (meetupDate == clickedDate) meetup else null
                }

                if (markedDates.isNotEmpty()) {
                    showMeetupDialog(markedDates)
                } else {
                    if(arguments?.getBoolean("fromLogin", false) != true){
                        start = clickedCalendar
                        end = clickedCalendar
                        showMeetupAlert()
                    } else{
                        fragmentEnterpriseCalendarBinding.calendarView.selectedDates = eventDates
                    }
                }
                if (!selectedDates.contains(clickedCalendar)) {
                    selectedDates.add(clickedCalendar)
                } else {
                    selectedDates.remove(clickedCalendar)
                }
            }
        })
        refreshCalendarView()
    }

    override fun onNewsItemClick(news: RealmNews?) {}
    override fun clearImages() {
        imageList.clear()
        llImage?.removeAllViews()
    }

    private fun getCardViewHeight(context: Context): Int {
        val view = LayoutInflater.from(context).inflate(R.layout.item_meetup, null)
        view.measure(
            View.MeasureSpec.makeMeasureSpec(Resources.getSystem().displayMetrics.widthPixels, View.MeasureSpec.AT_MOST),
            View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        )
        return view.measuredHeight
    }

    private fun showMeetupDialog(meetupList: List<RealmMeetup>) {
        val dialogView = LayoutInflater.from(requireContext()).inflate(R.layout.meetup_dialog, null)
        val recyclerView = dialogView.findViewById<RecyclerView>(R.id.rvMeetups)
        val dialogTitle = dialogView.findViewById< TextView>(R.id.tvTitle)
        val dateFormat = SimpleDateFormat("EEE, MMM d, yyyy", Locale.getDefault())
        dialogTitle.text = dateFormat.format(clickedCalendar.time)
        val extraHeight = TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP, 12f, resources.displayMetrics
        ).toInt()
        val cardHeight = getCardViewHeight(requireContext())
        recyclerView.layoutParams.height = cardHeight + extraHeight
        recyclerView.requestLayout()
        recyclerView.layoutManager = LinearLayoutManager(requireContext())
        recyclerView.adapter = AdapterMeetup(meetupList)

        val dialog = AlertDialog.Builder(requireContext())
            .setView(dialogView)
            .create()
        val btnAdd = dialogView.findViewById<Button>(R.id.btnadd)
        if (arguments?.getBoolean("fromLogin", false) != true) {
            btnAdd.visibility = View.VISIBLE
        } else {
            btnAdd.visibility = View.GONE
        }
        dialogView.findViewById<Button>(R.id.btnClose).setOnClickListener {
            dialog.dismiss()
            eventDates.add(clickedCalendar)
            fragmentEnterpriseCalendarBinding.calendarView.selectedDates = eventDates
        }
        btnAdd.setOnClickListener {
            if(arguments?.getBoolean("fromLogin", false) != true){
                start = clickedCalendar
                end = clickedCalendar
                showMeetupAlert()
            }
        }

        dialog.show()
    }

    private fun refreshCalendarView() {
        if (teamId.isEmpty()) {
            return
        }
        viewLifecycleOwner.lifecycleScope.launch(Dispatchers.IO) {
            var meetupList = mutableListOf<RealmMeetup>()
            val realm = Realm.getDefaultInstance()
            try {
                meetupList = realm.where(RealmMeetup::class.java).equalTo("teamId", teamId).findAll()
                val calendarInstance = Calendar.getInstance()

                for (meetup in meetupList) {
                    val startDateMillis = meetup.startDate
                    calendarInstance.timeInMillis = startDateMillis
                    eventDates.add(calendarInstance.clone() as Calendar)
                }
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                realm.close()
            }
            withContext(Dispatchers.Main) {
                if (isAdded && activity != null) {
                    fragmentEnterpriseCalendarBinding.calendarView.selectedDates = eventDates
                }
            }
        }
    }
}
